File: Combination_Sum_Backtracking.java

// Time Complexity : O(2^N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this : No
import java.util.*;

public class Combination_Sum_Backtracking {
	List<List<Integer>> result;
	int tar;

	public List<List<Integer>> combinationSum(int[] candidates, int target) {
		result = new ArrayList<>();
		tar = target;
		if (candidates == null || candidates.length == 0)
			return result;
		backtrack(candidates, new ArrayList<>(), 0, 0);
		return result;
	}

	private void backtrack(int candidates[], List<Integer> list, int Sum, int Index) {
		// base
		if (Sum > tar)
			return;
		if (Sum == tar) {
			result.add(new ArrayList<>(list));
			return;
		}
		// logic
		for (int i = Index; i < candidates.length; i++) {
			// action
			list.add(candidates[i]);
			// recurse
			backtrack(candidates, list, Sum + candidates[i], i);
			// backtrack
			list.remove(list.size() - 1);
		}
	}
}

File: combination-sum-backtracking.cpp
// Time Complexity : O(2^n+m)
// Space Complexity : O(2^n+m)
// Did this code successfully run on Leetcode : Yes

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> path;
        helper(candidates, 0, target, path, result);
        return result;
    }
private:
    void helper(vector<int>& candidates, int idx, int target, vector<int>& path, vector<vector<int>>& result){
        //base
        if(idx == candidates.size() || target < 0) return;
        if(target == 0){
            result.push_back(path);
            return;
        }
        //logic
        //not choose
        helper(candidates, idx + 1, target, path, result);

        //choose
        //action
        path.push_back(candidates[idx]);
        //recurse
        helper(candidates, idx, target - candidates[idx], path, result);
        //backtrack
        path.pop_back();
    }
};